src/main/scala/com/actian/spark_vector/vector/VectorJDBC.scala

```diff
@@ -15,12 +15,49 @@
  */
 package com.actian.spark_vector.vector
 
-import java.sql._
+import java.sql.{ Connection, DriverManager, PreparedStatement, ResultSet, Statement }
 import java.util.Properties
-import com.actian.spark_vector.util.ResourceUtil
-import com.actian.spark_vector.vector.ErrorCodes._
+
 import org.apache.spark.Logging
-import com.actian.spark_vector.sql.VectorRelation
+import org.apache.spark.sql.Row
+
+import com.actian.spark_vector.Profiling
+import com.actian.spark_vector.util.ResourceUtil.RichExtractableManagedResource
+import com.actian.spark_vector.vector.ErrorCodes.{ InvalidDataType, NoSuchTable, SqlException, SqlExecutionError }
+
+import resource.managed
+
+/** Iterator over an ResultSet */
+abstract class ResultSetIterator[T](result: ResultSet) extends Iterator[T] {
+  def extractor: ResultSet => T
+  override def hasNext: Boolean = result.next()
+  override def next(): T = extractor(result)
+}
+
+object ResultSetIterator {
+  def apply[T](result: ResultSet)(ex: ResultSet => T): ResultSetIterator[T] = new ResultSetIterator[T](result) {
+    def extractor = ex
+  }
+}
+
+/** Extracts `Rows` out of a `ResultSet` */
+class ResultSetToRowIterator(result: ResultSet) extends ResultSetIterator[Row](result) with Logging with Profiling {
+  lazy val numColumns = result.getMetaData.getColumnCount
+  lazy val row = collection.mutable.IndexedSeq.fill[Any](numColumns)(null)
+  implicit lazy val prof_accs = profileInit("resultSet extraction")
+  var col = 0
+  def extractor: ResultSet => Row = { rs =>
+    profile("resultSet extraction")
+    col = 0
+    while (col < numColumns) {
```

Comment:
for (col <- 0 to numColumns - 1) {
}

 // or for (col <- 0 until numColumns), though I prefer the explicit -1 version